rust/repositories.bzl

```diff
@@ -42,98 +55,272 @@ filegroup(
+    """Verifies that the provided rust version and iso_date make sense."""
+
+    if not version and iso_date:
+        fail("{param_prefix}iso_date must be paired with a {param_prefix}version".format(param_prefix = param_prefix))
+
+    if version in ("beta", "nightly") and not iso_date:
+        fail("{param_prefix}iso_date must be specified if version is 'beta' or 'nightly'".format(param_prefix = param_prefix))
+
+    if version not in ("beta", "nightly") and iso_date:
+        print("{param_prefix}iso_date is ineffective if an exact version is specified".format(param_prefix = param_prefix))
+
+def produce_tool_suburl(tool_name, target_triple, version, iso_date = None):
+    """Produces a fully qualified Rust tool name for URL
+
+    Args:
+      tool_name: The name of the tool per static.rust-lang.org
+      target_triple: The rust-style target triple of the tool
+      version: The version of the tool among "nightly", "beta', or an exact version.
+      iso_date: The date of the tool (or None, if the version is a specific version).
+    """
+
+    if iso_date:
+        return "{}/{}-{}-{}".format(iso_date, tool_name, version, target_triple)
+    else:
+        return "{}-{}-{}".format(tool_name, version, target_triple)
+
+def produce_tool_path(tool_name, target_triple, version):
+    """Produces a qualified Rust tool name
+
+    Args:
+      tool_name: The name of the tool per static.rust-lang.org
+      target_triple: The rust-style target triple of the tool
+      version: The version of the tool among "nightly", "beta', or an exact version.
+    """
+
+    return "{}-{}-{}".format(tool_name, version, target_triple)
+
+def load_arbitrary_tool(ctx, tool_name, param_prefix, tool_subdirectory, version, iso_date, target_triple):
+    """Loads a Rust tool, downloads, and extracts into the common workspace.
+
+    Args:
+      ctx: A repository_ctx.
+      tool_name: The name of the given tool per the archive naming.
+      param_prefix: The name of the versioning param if the repository rule supports multiple tools.
+      tool_subdirectory: The subdirectory of the tool files (wo level below the root directory of
+                         the archive. The root directory of the archive is expected to match
+                         $TOOL_NAME-$VERSION-$TARGET_TRIPLE.
+      version: The version of the tool among "nightly", "beta', or an exact version.
+      iso_date: The date of the tool (or None, if the version is a specific version).
+      target_triple: The rust-style target triple of the tool
+    """
+
+    _check_version_valid(version, iso_date, param_prefix)
+
+    tool_suburl = produce_tool_suburl(tool_name, target_triple, version, iso_date)
+    url = "https://static.rust-lang.org/dist/{}.tar.gz".format(tool_suburl)
+
+    tool_path = produce_tool_path(tool_name, target_triple, version)
+    ctx.download_and_extract(
+        url,
+        output = "",
+        sha256 = FILE_KEY_TO_SHA.get(tool_suburl) or "",
```

Comment:
This basically means users can't use versions rules_rust doesn't know about without patching it themselves (unless providing "" skips the check?)

Also this doesn't work well with vendored toolchains that would've been setup w/ new_local_repository.. not sure what to do here.

Reply:
Empty string does indeed skip the sha check. 

I do want to have a good story for reusing these pieces with a vendored toolchain. I'm thinking we have that already (with some assembly required), but I'm ready for more feedback. Here's what I have in mind:

Assuming a toolchain is prepared to match a rustup-style toolchain (e.g. you can use rustup dir itself)

//:toolchain.bzl
```python
def rustup_derived_toolchain(name):
  new_local_reposistory(
    name = name,
    path = "/home/acmcarther/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/"
    build_file_content = "\n".join([
      BUILD_for_compiler("x86_64-unknown-linux-gnu"),
      BUILD_for_stdlib("x86_64-unknown-linux-gnu"),
      BUILD_for_toolchain(
        workspace_name = name, 
        name = "toolchain_for_x86_64-unknown-linux-gnu", 
        exec_triple = "x86_64-unknown-linux-gnu",
        target_triple = "x86_64-unknown-linux-gnu"
      )]
    )
```

WORKSPACE
```python
load(":toolchain.bzl", "rustup_derived_toolchain")

rustup_derived_toolchain("rustup_toolchain")
```

Reply:
iirc the rustup layout is different than the tar, but that bit of finagling ought to be ok

Reply:
I don't want to belabor the point too much, but all of the filesystem shenanigans here are built around making our usage of the raw tars match rustup _exactly_, because the compiler has expectations baked into it that files live in certain places relative to each other. This wasn't a problem for us before because using the "rust" download rather than "rustc plus rust-std" incidentally puts rustc in the right place relative to the stdlib.

For clarity, I'll add the below to the PR description as well.

---
Example comparison between Rustup and this repository_rule:
- For Bazel: Added x86_64-unknown-freebsd to `extra_target_triples`
- For Rustup: Ran `rustup target add x86_64-unknown-freebsd`

Dir structure comparison
[$tree -I "share" external/rust_linux_x86_64](https://gist.github.com/acmcarther/6e3f8b439aacb6760bea975311de2209)
[$tree -I "share" ~/.rustup/toolchains/stable-x86_64-unknown-linux-gnu"](https://gist.github.com/acmcarther/b20348cc6f93cfaf1d640045109dec0e)